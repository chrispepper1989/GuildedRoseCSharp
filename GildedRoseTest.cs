using System;
using System.IO;
using System.Text;
using ApprovalTests;
using ApprovalTests.Reporters;
using csharp;

namespace GildedRoseCSharp
{
  
    [UseReporter(typeof(DiffReporter))]
    [TestFixture]
    public class GildedRoseTest
    {

        //generate a bunch of test cases
        public IList<Item> GenerateTestItems(ValidItems[] names, int[] sellIns, int[] qualities)
        {
            IList<Item> Items = new List<Item>
            {
            };
            foreach (var name in names)
            {
                foreach (var sellIn in sellIns)
                {
                    foreach (var quality in qualities)
                    {
                        var item = ItemHelper.CreateItem(name, sellIn, quality);
                        Items.Add( item);
                    }

                }

            }

            return Items;

        }

        //[TestCase(10)] //todo why doesnt this work with Approvals?
        public void TestUpdateQuality(int days)
        {

            //arrange
            ValidItems[] namesProper = new[]
            {
                ValidItems.AgedBrie, ValidItems.DexterityVest, ValidItems.ElixirOfMongoose,
                ValidItems.SulfurasHandOfRagnaros, ValidItems.ConjuredManaCake
            };
            var sellIns = new[] { 1, 5, 10, 2, 4 };
            var qualities = new[] { 0, 1, 2, 5, 29 };
            var Items = GenerateTestItems(namesProper, sellIns, qualities);
            GildedRose app = new GildedRose(Items);
            //act
            for (var i = 0; i < days; ++i)
            {
                app.UpdateQuality();
            }
            //assert
            Approvals.VerifyAll(Items, "All Items");

        }
        [Test]
        public void ConjuredManaCakeDegradesTwiceAsFastAsNormalItems()
        {
            //arrange
            var conitem = ItemHelper.CreateItem(ValidItems.ConjuredManaCake, 10, 100);
            var normalItem = ItemHelper.CreateItem(ValidItems.ElixirOfMongoose, 10, 100);

            IList<Item> Items = new List<Item>
            {
         
            };

            GildedRose app = new GildedRose(Items);

            IList<Item> ItemsOverTime = new List<Item>();
            //act
            for (var i = 0; i < 100; ++i)
            {
                conitem.SellIn -= 1;
                normalItem.SellIn -= 1;
                //normal item done twice
                normalItem.Quality =  GildedRose.GetNewStandardItemQuality(normalItem);
                normalItem.Quality = GildedRose.GetNewStandardItemQuality(normalItem);
                conitem.Quality = GildedRose.GetNewConjuredManaCakeQuality(conitem);
               
                //assert
                Assert.AreEqual(normalItem.Quality, conitem.Quality);
      
                ItemsOverTime.Add(new Item()
                {
                    Name = normalItem.Name,
                    Quality = normalItem.Quality,
                    SellIn = normalItem.SellIn

                });
                ItemsOverTime.Add(new Item()
                {
                    Name = conitem.Name,
                    Quality = conitem.Quality,
                    SellIn = conitem.SellIn

                });

            }
            Approvals.VerifyAll(ItemsOverTime, "Conjured Items ");

        }

        [Test]
         public void TestUpdateQuality_1Days()
         {
             TestUpdateQuality(1);
         }
         [Test]
         public void TestUpdateQuality_5Days()
         {
             TestUpdateQuality(5);
         }
         [Test]
         public void TestUpdateQuality_10Days()
         {
             TestUpdateQuality(10);
         }
         [Test]
         public void TestUpdateQuality_30Days()
         {
             TestUpdateQuality(30);
         }

    }
}
