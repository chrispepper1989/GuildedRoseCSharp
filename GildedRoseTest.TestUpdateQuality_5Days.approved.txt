All Items[0] = Aged Brie, -4, 9
All Items[1] = Aged Brie, -4, 10
All Items[2] = Aged Brie, -4, 11
All Items[3] = Aged Brie, -4, 14
All Items[4] = Aged Brie, -4, 38
All Items[5] = Aged Brie, -4, 39
All Items[6] = Aged Brie, -4, 50
All Items[7] = Aged Brie, -4, 75
All Items[8] = Aged Brie, -4, 90
All Items[9] = Aged Brie, -4, 100
All Items[10] = Aged Brie, 0, 5
All Items[11] = Aged Brie, 0, 6
All Items[12] = Aged Brie, 0, 7
All Items[13] = Aged Brie, 0, 10
All Items[14] = Aged Brie, 0, 34
All Items[15] = Aged Brie, 0, 35
All Items[16] = Aged Brie, 0, 50
All Items[17] = Aged Brie, 0, 75
All Items[18] = Aged Brie, 0, 90
All Items[19] = Aged Brie, 0, 100
All Items[20] = Aged Brie, 5, 5
All Items[21] = Aged Brie, 5, 6
All Items[22] = Aged Brie, 5, 7
All Items[23] = Aged Brie, 5, 10
All Items[24] = Aged Brie, 5, 34
All Items[25] = Aged Brie, 5, 35
All Items[26] = Aged Brie, 5, 50
All Items[27] = Aged Brie, 5, 75
All Items[28] = Aged Brie, 5, 90
All Items[29] = Aged Brie, 5, 100
All Items[30] = Aged Brie, -3, 8
All Items[31] = Aged Brie, -3, 9
All Items[32] = Aged Brie, -3, 10
All Items[33] = Aged Brie, -3, 13
All Items[34] = Aged Brie, -3, 37
All Items[35] = Aged Brie, -3, 38
All Items[36] = Aged Brie, -3, 50
All Items[37] = Aged Brie, -3, 75
All Items[38] = Aged Brie, -3, 90
All Items[39] = Aged Brie, -3, 100
All Items[40] = Aged Brie, -1, 6
All Items[41] = Aged Brie, -1, 7
All Items[42] = Aged Brie, -1, 8
All Items[43] = Aged Brie, -1, 11
All Items[44] = Aged Brie, -1, 35
All Items[45] = Aged Brie, -1, 36
All Items[46] = Aged Brie, -1, 50
All Items[47] = Aged Brie, -1, 75
All Items[48] = Aged Brie, -1, 90
All Items[49] = Aged Brie, -1, 100
All Items[50] = +5 Dexterity Vest, -4, 0
All Items[51] = +5 Dexterity Vest, -4, 0
All Items[52] = +5 Dexterity Vest, -4, 0
All Items[53] = +5 Dexterity Vest, -4, 0
All Items[54] = +5 Dexterity Vest, -4, 20
All Items[55] = +5 Dexterity Vest, -4, 21
All Items[56] = +5 Dexterity Vest, -4, 41
All Items[57] = +5 Dexterity Vest, -4, 66
All Items[58] = +5 Dexterity Vest, -4, 81
All Items[59] = +5 Dexterity Vest, -4, 91
All Items[60] = +5 Dexterity Vest, 0, 0
All Items[61] = +5 Dexterity Vest, 0, 0
All Items[62] = +5 Dexterity Vest, 0, 0
All Items[63] = +5 Dexterity Vest, 0, 0
All Items[64] = +5 Dexterity Vest, 0, 24
All Items[65] = +5 Dexterity Vest, 0, 25
All Items[66] = +5 Dexterity Vest, 0, 45
All Items[67] = +5 Dexterity Vest, 0, 70
All Items[68] = +5 Dexterity Vest, 0, 85
All Items[69] = +5 Dexterity Vest, 0, 95
All Items[70] = +5 Dexterity Vest, 5, 0
All Items[71] = +5 Dexterity Vest, 5, 0
All Items[72] = +5 Dexterity Vest, 5, 0
All Items[73] = +5 Dexterity Vest, 5, 0
All Items[74] = +5 Dexterity Vest, 5, 24
All Items[75] = +5 Dexterity Vest, 5, 25
All Items[76] = +5 Dexterity Vest, 5, 45
All Items[77] = +5 Dexterity Vest, 5, 70
All Items[78] = +5 Dexterity Vest, 5, 85
All Items[79] = +5 Dexterity Vest, 5, 95
All Items[80] = +5 Dexterity Vest, -3, 0
All Items[81] = +5 Dexterity Vest, -3, 0
All Items[82] = +5 Dexterity Vest, -3, 0
All Items[83] = +5 Dexterity Vest, -3, 0
All Items[84] = +5 Dexterity Vest, -3, 21
All Items[85] = +5 Dexterity Vest, -3, 22
All Items[86] = +5 Dexterity Vest, -3, 42
All Items[87] = +5 Dexterity Vest, -3, 67
All Items[88] = +5 Dexterity Vest, -3, 82
All Items[89] = +5 Dexterity Vest, -3, 92
All Items[90] = +5 Dexterity Vest, -1, 0
All Items[91] = +5 Dexterity Vest, -1, 0
All Items[92] = +5 Dexterity Vest, -1, 0
All Items[93] = +5 Dexterity Vest, -1, 0
All Items[94] = +5 Dexterity Vest, -1, 23
All Items[95] = +5 Dexterity Vest, -1, 24
All Items[96] = +5 Dexterity Vest, -1, 44
All Items[97] = +5 Dexterity Vest, -1, 69
All Items[98] = +5 Dexterity Vest, -1, 84
All Items[99] = +5 Dexterity Vest, -1, 94
All Items[100] = Elixir of the Mongoose, -4, 0
All Items[101] = Elixir of the Mongoose, -4, 0
All Items[102] = Elixir of the Mongoose, -4, 0
All Items[103] = Elixir of the Mongoose, -4, 0
All Items[104] = Elixir of the Mongoose, -4, 20
All Items[105] = Elixir of the Mongoose, -4, 21
All Items[106] = Elixir of the Mongoose, -4, 41
All Items[107] = Elixir of the Mongoose, -4, 66
All Items[108] = Elixir of the Mongoose, -4, 81
All Items[109] = Elixir of the Mongoose, -4, 91
All Items[110] = Elixir of the Mongoose, 0, 0
All Items[111] = Elixir of the Mongoose, 0, 0
All Items[112] = Elixir of the Mongoose, 0, 0
All Items[113] = Elixir of the Mongoose, 0, 0
All Items[114] = Elixir of the Mongoose, 0, 24
All Items[115] = Elixir of the Mongoose, 0, 25
All Items[116] = Elixir of the Mongoose, 0, 45
All Items[117] = Elixir of the Mongoose, 0, 70
All Items[118] = Elixir of the Mongoose, 0, 85
All Items[119] = Elixir of the Mongoose, 0, 95
All Items[120] = Elixir of the Mongoose, 5, 0
All Items[121] = Elixir of the Mongoose, 5, 0
All Items[122] = Elixir of the Mongoose, 5, 0
All Items[123] = Elixir of the Mongoose, 5, 0
All Items[124] = Elixir of the Mongoose, 5, 24
All Items[125] = Elixir of the Mongoose, 5, 25
All Items[126] = Elixir of the Mongoose, 5, 45
All Items[127] = Elixir of the Mongoose, 5, 70
All Items[128] = Elixir of the Mongoose, 5, 85
All Items[129] = Elixir of the Mongoose, 5, 95
All Items[130] = Elixir of the Mongoose, -3, 0
All Items[131] = Elixir of the Mongoose, -3, 0
All Items[132] = Elixir of the Mongoose, -3, 0
All Items[133] = Elixir of the Mongoose, -3, 0
All Items[134] = Elixir of the Mongoose, -3, 21
All Items[135] = Elixir of the Mongoose, -3, 22
All Items[136] = Elixir of the Mongoose, -3, 42
All Items[137] = Elixir of the Mongoose, -3, 67
All Items[138] = Elixir of the Mongoose, -3, 82
All Items[139] = Elixir of the Mongoose, -3, 92
All Items[140] = Elixir of the Mongoose, -1, 0
All Items[141] = Elixir of the Mongoose, -1, 0
All Items[142] = Elixir of the Mongoose, -1, 0
All Items[143] = Elixir of the Mongoose, -1, 0
All Items[144] = Elixir of the Mongoose, -1, 23
All Items[145] = Elixir of the Mongoose, -1, 24
All Items[146] = Elixir of the Mongoose, -1, 44
All Items[147] = Elixir of the Mongoose, -1, 69
All Items[148] = Elixir of the Mongoose, -1, 84
All Items[149] = Elixir of the Mongoose, -1, 94
All Items[150] = Sulfuras, Hand of Ragnaros, 1, 0
All Items[151] = Sulfuras, Hand of Ragnaros, 1, 1
All Items[152] = Sulfuras, Hand of Ragnaros, 1, 2
All Items[153] = Sulfuras, Hand of Ragnaros, 1, 5
All Items[154] = Sulfuras, Hand of Ragnaros, 1, 29
All Items[155] = Sulfuras, Hand of Ragnaros, 1, 30
All Items[156] = Sulfuras, Hand of Ragnaros, 1, 50
All Items[157] = Sulfuras, Hand of Ragnaros, 1, 75
All Items[158] = Sulfuras, Hand of Ragnaros, 1, 90
All Items[159] = Sulfuras, Hand of Ragnaros, 1, 100
All Items[160] = Sulfuras, Hand of Ragnaros, 5, 0
All Items[161] = Sulfuras, Hand of Ragnaros, 5, 1
All Items[162] = Sulfuras, Hand of Ragnaros, 5, 2
All Items[163] = Sulfuras, Hand of Ragnaros, 5, 5
All Items[164] = Sulfuras, Hand of Ragnaros, 5, 29
All Items[165] = Sulfuras, Hand of Ragnaros, 5, 30
All Items[166] = Sulfuras, Hand of Ragnaros, 5, 50
All Items[167] = Sulfuras, Hand of Ragnaros, 5, 75
All Items[168] = Sulfuras, Hand of Ragnaros, 5, 90
All Items[169] = Sulfuras, Hand of Ragnaros, 5, 100
All Items[170] = Sulfuras, Hand of Ragnaros, 10, 0
All Items[171] = Sulfuras, Hand of Ragnaros, 10, 1
All Items[172] = Sulfuras, Hand of Ragnaros, 10, 2
All Items[173] = Sulfuras, Hand of Ragnaros, 10, 5
All Items[174] = Sulfuras, Hand of Ragnaros, 10, 29
All Items[175] = Sulfuras, Hand of Ragnaros, 10, 30
All Items[176] = Sulfuras, Hand of Ragnaros, 10, 50
All Items[177] = Sulfuras, Hand of Ragnaros, 10, 75
All Items[178] = Sulfuras, Hand of Ragnaros, 10, 90
All Items[179] = Sulfuras, Hand of Ragnaros, 10, 100
All Items[180] = Sulfuras, Hand of Ragnaros, 2, 0
All Items[181] = Sulfuras, Hand of Ragnaros, 2, 1
All Items[182] = Sulfuras, Hand of Ragnaros, 2, 2
All Items[183] = Sulfuras, Hand of Ragnaros, 2, 5
All Items[184] = Sulfuras, Hand of Ragnaros, 2, 29
All Items[185] = Sulfuras, Hand of Ragnaros, 2, 30
All Items[186] = Sulfuras, Hand of Ragnaros, 2, 50
All Items[187] = Sulfuras, Hand of Ragnaros, 2, 75
All Items[188] = Sulfuras, Hand of Ragnaros, 2, 90
All Items[189] = Sulfuras, Hand of Ragnaros, 2, 100
All Items[190] = Sulfuras, Hand of Ragnaros, 4, 0
All Items[191] = Sulfuras, Hand of Ragnaros, 4, 1
All Items[192] = Sulfuras, Hand of Ragnaros, 4, 2
All Items[193] = Sulfuras, Hand of Ragnaros, 4, 5
All Items[194] = Sulfuras, Hand of Ragnaros, 4, 29
All Items[195] = Sulfuras, Hand of Ragnaros, 4, 30
All Items[196] = Sulfuras, Hand of Ragnaros, 4, 50
All Items[197] = Sulfuras, Hand of Ragnaros, 4, 75
All Items[198] = Sulfuras, Hand of Ragnaros, 4, 90
All Items[199] = Sulfuras, Hand of Ragnaros, 4, 100
All Items[200] = Sulfuras, Hand of Ragnaros, 1, 0
All Items[201] = Sulfuras, Hand of Ragnaros, 1, 1
All Items[202] = Sulfuras, Hand of Ragnaros, 1, 2
All Items[203] = Sulfuras, Hand of Ragnaros, 1, 5
All Items[204] = Sulfuras, Hand of Ragnaros, 1, 29
All Items[205] = Sulfuras, Hand of Ragnaros, 1, 30
All Items[206] = Sulfuras, Hand of Ragnaros, 1, 50
All Items[207] = Sulfuras, Hand of Ragnaros, 1, 75
All Items[208] = Sulfuras, Hand of Ragnaros, 1, 90
All Items[209] = Sulfuras, Hand of Ragnaros, 1, 100
All Items[210] = Sulfuras, Hand of Ragnaros, 5, 0
All Items[211] = Sulfuras, Hand of Ragnaros, 5, 1
All Items[212] = Sulfuras, Hand of Ragnaros, 5, 2
All Items[213] = Sulfuras, Hand of Ragnaros, 5, 5
All Items[214] = Sulfuras, Hand of Ragnaros, 5, 29
All Items[215] = Sulfuras, Hand of Ragnaros, 5, 30
All Items[216] = Sulfuras, Hand of Ragnaros, 5, 50
All Items[217] = Sulfuras, Hand of Ragnaros, 5, 75
All Items[218] = Sulfuras, Hand of Ragnaros, 5, 90
All Items[219] = Sulfuras, Hand of Ragnaros, 5, 100
All Items[220] = Sulfuras, Hand of Ragnaros, 10, 0
All Items[221] = Sulfuras, Hand of Ragnaros, 10, 1
All Items[222] = Sulfuras, Hand of Ragnaros, 10, 2
All Items[223] = Sulfuras, Hand of Ragnaros, 10, 5
All Items[224] = Sulfuras, Hand of Ragnaros, 10, 29
All Items[225] = Sulfuras, Hand of Ragnaros, 10, 30
All Items[226] = Sulfuras, Hand of Ragnaros, 10, 50
All Items[227] = Sulfuras, Hand of Ragnaros, 10, 75
All Items[228] = Sulfuras, Hand of Ragnaros, 10, 90
All Items[229] = Sulfuras, Hand of Ragnaros, 10, 100
All Items[230] = Sulfuras, Hand of Ragnaros, 2, 0
All Items[231] = Sulfuras, Hand of Ragnaros, 2, 1
All Items[232] = Sulfuras, Hand of Ragnaros, 2, 2
All Items[233] = Sulfuras, Hand of Ragnaros, 2, 5
All Items[234] = Sulfuras, Hand of Ragnaros, 2, 29
All Items[235] = Sulfuras, Hand of Ragnaros, 2, 30
All Items[236] = Sulfuras, Hand of Ragnaros, 2, 50
All Items[237] = Sulfuras, Hand of Ragnaros, 2, 75
All Items[238] = Sulfuras, Hand of Ragnaros, 2, 90
All Items[239] = Sulfuras, Hand of Ragnaros, 2, 100
All Items[240] = Sulfuras, Hand of Ragnaros, 4, 0
All Items[241] = Sulfuras, Hand of Ragnaros, 4, 1
All Items[242] = Sulfuras, Hand of Ragnaros, 4, 2
All Items[243] = Sulfuras, Hand of Ragnaros, 4, 5
All Items[244] = Sulfuras, Hand of Ragnaros, 4, 29
All Items[245] = Sulfuras, Hand of Ragnaros, 4, 30
All Items[246] = Sulfuras, Hand of Ragnaros, 4, 50
All Items[247] = Sulfuras, Hand of Ragnaros, 4, 75
All Items[248] = Sulfuras, Hand of Ragnaros, 4, 90
All Items[249] = Sulfuras, Hand of Ragnaros, 4, 100
